server/chat-commands/moderation.ts

```diff
@@ -102,52 +100,56 @@ exports.commands = {
 			return this.errorReply(`Group 'room${Config.groups[nextGroup].id}' does not exist as a room rank.`);
 		}
 
-		let groupName = Config.groups[nextGroup].name || "regular user";
-		if ((room.auth[userid] || Config.groupsranking[0]) === nextGroup) {
+		const groupName = Config.groups[nextGroup].name || "regular user";
+		if (room.auth && (room.auth[userid] || Config.groupsranking[0]) === nextGroup) {
 			return this.errorReply(`User '${name}' is already a ${groupName} in this room.`);
 		}
 		if (!user.can('makeroom')) {
-			if (currentGroup !== ' ' && !user.can('room' + (Config.groups[currentGroup] ? Config.groups[currentGroup].id : 'voice'), null, room)) {
+			if (currentGroup !== ' ' &&
+			!user.can('room' + (Config.groups[currentGroup] ? Config.groups[currentGroup].id : 'voice'), null, room)) {
 				return this.errorReply(`/${cmd} - Access denied for promoting/demoting from ${(Config.groups[currentGroup] ? Config.groups[currentGroup].name : "an undefined group")}.`);
 			}
 			if (nextGroup !== ' ' && !user.can('room' + Config.groups[nextGroup].id, null, room)) {
 				return this.errorReply(`/${cmd} - Access denied for promoting/demoting to ${Config.groups[nextGroup].name}.`);
 			}
 		}
-		let nextGroupIndex = Config.groupsranking.indexOf(nextGroup) || 1; // assume voice if not defined (although it should be by now)
-		if (targetUser && targetUser.locked && !room.isPrivate && !room.battle && !room.isPersonal && nextGroupIndex >= 2) {
+		const nextGroupIndex = Config.groupsranking.indexOf(nextGroup) || 1;
+		// assume voice if not defined (although it should be by now)
+		if (targetUser?.locked && !room.isPrivate && !room.battle && !room.isPersonal && nextGroupIndex >= 2) {
 			return this.errorReply("Locked users can't be promoted.");
 		}
 
 		if (nextGroup === Config.groupsranking[0]) {
-			delete room.auth[userid];
+			if (room.auth) delete room.auth[userid];
 		} else {
-			room.auth[userid] = nextGroup;
+			if (room.auth) room.auth[userid] = nextGroup as GroupSymbol;
 		}
 
 		// Only show popup if: user is online and in the room, the room is public, and not a groupchat or a battle.
-		let needsPopup = targetUser && room.users[targetUser.id] && !room.isPrivate && !room.isPersonal && !room.battle;
+		const shouldPopup = (room.users[targetUser!.id] && !room.isPrivate &&
+			!room.isPersonal && !room.battle ? targetUser : null);
```

Comment:
```suggestion
		const shouldPopup = (
			targetUser && room.users[targetUser.id] && !room.isPrivate && !room.isPersonal && !room.battle ?
			targetUser : null
		);
```

you can't do that.......